paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/ProjectionRowData.java

```diff
@@ -0,0 +1,199 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.DecimalData;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.data.RawValueData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.types.RowKind;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+
+/**
+ * A {@link RowData} that provides a mapping view of the original {@link RowData} according to
+ * projection information.
+ */
+public class ProjectionRowData implements RowData, Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final RowType producedDataType;
+    private final int[][] projectedFields;
+    private final FieldGetter[][] fieldGetters;
+
+    private transient GenericRowData row;
+
+    ProjectionRowData(RowType producedDataType, int[][] projectedFields) {
+        this.producedDataType = producedDataType;
+        this.projectedFields = projectedFields;
+        this.fieldGetters = new FieldGetter[projectedFields.length][];
+        for (int i = 0; i < projectedFields.length; i++) {
+            this.fieldGetters[i] = new FieldGetter[projectedFields[i].length];
+            LogicalType currentType = producedDataType;
+            for (int j = 0; j < projectedFields[i].length; j++) {
+                currentType = ((RowType) currentType).getTypeAt(projectedFields[i][j]);
+                this.fieldGetters[i][j] =
+                        RowData.createFieldGetter(currentType, projectedFields[i][j]);
+            }
+        }
+    }
+
+    public ProjectionRowData replaceRow(RowData inputRow) {
+        if (this.row == null) {
+            this.row = new GenericRowData(inputRow.getRowKind(), fieldGetters.length);
+        }
+
+        for (int i = 0; i < fieldGetters.length; i++) {
+            Object currentRow = inputRow;
+            for (int j = 0; j < fieldGetters[i].length; j++) {
+                if (currentRow == null) {
+                    break;
+                }
+                currentRow = this.fieldGetters[i][j].getFieldOrNull((RowData) currentRow);
+            }
+            this.row.setField(i, currentRow);
+        }
+
+        if (inputRow != null) {
+            this.row.setRowKind(inputRow.getRowKind());
+        }
+        return this;
+    }
```

Comment:
So you extract all fields from `inputRow` in this method? This is too heavy. See implementation of `ProjectedRow`, and only fetch fields when required.